src/tox_ansible/plugin.py

```diff
@@ -0,0 +1,528 @@
+    )
+
+    parser.add_argument(
+        "--ansible",
+        action="store_true",
+        default=False,
+        help="Enable ansible testing",
+    )
+
+
+@impl
+def tox_add_core_config(
+    core_conf: CoreConfigSet,  # pylint: disable=unused-argument
+    state: State,
+) -> None:
+    """Dump the environment list and exit.
+
+    :param core_conf: The core configuration object.
+    :param state: The state object.
+    """
+    if state.conf.options.gh_matrix and not state.conf.options.ansible:
+        err = "The --gh-matrix option requires --ansible"
+        logging.critical(err)
+        sys.exit(1)
+
+    if not state.conf.options.ansible:
+        return
+
+    global TOX_WORK_DIR  # pylint: disable=global-statement # noqa: PLW0603
+    TOX_WORK_DIR = state.conf.work_dir
+    env_list = add_ansible_matrix(state)
+
+    if not state.conf.options.gh_matrix:
+        return
+
+    generate_gh_matrix(env_list=env_list)
+    sys.exit(0)
+
+
+@impl
+def tox_add_env_config(env_conf: EnvConfigSet, state: State) -> None:
+    """Add the test requirements and ansible-core to the virtual environment.
+
+    :param env_conf: The environment configuration object.
+    :param state: The state object.
+    """
+    if not state.conf.options.ansible:
+        return
+
+    test_type = env_conf.name.split("-")[0]
+    if test_type not in ["integration", "sanity", "unit"]:
+        return
+
+    galaxy_path = TOX_WORK_DIR / "galaxy.yml"
+    c_name, c_namespace = get_collection_name(galaxy_path=galaxy_path)
+
+    conf = AnsibleTestConf(
+        allowlist_externals=ALLOWED_EXTERNALS,
+        commands_pre=conf_commands_pre(
+            c_name=c_name,
+            c_namespace=c_namespace,
+            env_conf=env_conf,
+        ),
+        commands=conf_commands(
+            c_name=c_name,
+            c_namespace=c_namespace,
+            env_conf=env_conf,
+            test_type=test_type,
+        ),
+        deps=conf_deps(env_conf=env_conf, test_type=test_type),
+        passenv=conf_passenv(),
+        setenv=conf_setenv(env_conf),
+        skip_install=True,
+    )
+    # Add any user provided configuration
+    u_allowlist_externals = env_conf["allowlist_externals"]
+    u_commands_pre = [cmd.shell for cmd in env_conf["commands_pre"]]
+    u_commands = [cmd.shell for cmd in env_conf["commands"]]
+    u_deps = "\n".join(env_conf["deps"].as_root_args)
+    u_passenv = env_conf["pass_env"]
+    # pylint: disable=protected-access
+    u_setenv = "\n".join([f"{k}={v}" for k, v in env_conf["set_env"]._raw.items()])  # noqa: SLF001
+    # pylint: enable=protected-access
+
+    conf.allowlist_externals.extend(u_allowlist_externals)
+    conf.commands_pre.extend(u_commands_pre)
+    conf.commands.extend(u_commands)
+    conf.deps = f"{conf.deps}\n{u_deps}"
+    conf.passenv.extend(u_passenv)
+    conf.setenv = f"{conf.setenv}\n{u_setenv}"
+
+    loader = MemoryLoader(**asdict(conf))
+    env_conf.loaders.insert(0, loader)
```

Comment:
If you change THIS line, you can skip all the stuff above and just leave it alone. Then the user will be able to override the dfeaults in this plugin by manually entering them into the ini file. This is because the internals of tox iterate through the `env_conf.loaders` and query each entry for a value, then take the first value one of the loaders returns. When you have the line as currently written, then the plugin takes precedence over anything the user sets as a value. When you change it to my suggestion, the value in the ini file takes precedence, and the values of this plugin fill in default values.

You should revert the code above this back to its simple form of just providing the default values, then use the suggestion I'm putting here to allow INI file overrides.

```suggestion
    env_conf.loaders.append(loader)
```

Reply:
Given this sample INI file entry

```ini
[testenv:a]
commands = echo hi from tox.ini
```

And the following plugin code:

```python
env_conf.loaders.insert(0, MemoryLoader(commands=["echo hi"], deps=["ansible"]))
```

I get the following by running `tox -e a`:
`a: commands[0]> echo hi`

If I change the Python line to read
```python
env_conf.loaders.append(MemoryLoader(commands=["echo hi"], deps=["ansible"]))
```

I get the following by running `tox -e a`:
`a: commands[0]> echo hi from tox.ini`

Both of them still have Ansible in the list of deps, because I do not explicitly set any value for deps in the INI file. However, I am able to customize any portion of it I want, using standard Tox methods in the tox.ini file. Personally, I think this would be superior to appending to them or to overriding them with the previous code.